### ISSUE TITLE
Cannot write runtime configuration if release dir is read-only

### ISSUE THREAD

**Issue by goodsoft** [elixir-lang/elixir#9338]
### Environment

* Elixir & Erlang/OTP versions (elixir --version): 1.9.1, Erlang/OTP 22
* Operating system: NixOS

### Current behavior

We use native Elixir releases to distribute our application, and use `config/releases.exs` for its runtime configuration. As our build process is based on Nix, the release itself (i.e. `RELEASE_ROOT`) resides on read-only file system. We provide writable directory to the application using `RELEASE_TMP` environment variable.

The application fails:
```
ERROR! Could not write runtime configuration due to reason: :eacces. Please make sure "/tmp/app-0.1.0-20190910170908-8803.runtime.config" is writable and accessible or choose a different path
{"init terminating in do_boot",{#{'__exception__'=>true,'__struct__'=>'Elixir.RuntimeError',message=><<"Could not write runtime configuration due to reason: :eacces. Please make sure \"/tmp/app-0.1.0-20190910170908-8803.runtime.config\" is writable and accessible or choose a different path">>},[{'Elixir.Config.Provider',abort,1,[{file,"lib/config/provider.ex"},{line,247}]},{'Elixir.Config.Provider',boot,3,[{file,"lib/config/provider.ex"},{line,163}]},{init,eval_script,2,[]},{init,do_boot,3,[]}]}}
```

We figured out that this happens due to boot script using [`cp`](https://github.com/elixir-lang/elixir/blob/802461b4ec35b8058a847c2a53906660d91873ff/lib/mix/lib/mix/tasks/release.init.ex#L137) to copy `sys.config` from the release directory, which is, as mentioned above, read-only. The resulting file's read-only permissions are retained, and runtime configuration cannot be written to it.

### Expected behavior

Boot script should disregard original `sys.config` permissions and create a new writable file.

### Workaround

Put this line into `config/releases.exs`:
```
File.chmod(System.get_env("RELEASE_SYS_CONFIG") <> ".config", 0o644)
```

**Comment by josevalim** [comment]
Hi @goodsoft!

It does look like a bug but I am unable to reproduce it locally. Can you please provide the steps to reproduce it? I want to reproduce it so I can be sure I am indeed fixing it. Here is what I tried:

```
$ mix new omg
$ cd omg
$ mkdir config
$ echo "import Config" > config/releases.exs
$ mix release
$ chmod a=rw _build/dev/rel/omg/releases/0.1.0
$ _build/dev/rel/omg/bin/omg start
```

In my case the app booted just fine, which makes me think there is something else at play here. Does the above work for you too?

**Comment by yorickvP** [comment]
`chmod a=rw` will set read-write permissions. You will need `chmod a=rx` on the release directory.

**Comment by josevalim** [comment]
@yorickvP doh, i meant to set rx but I guess my brain is not fully functional today. Thank you for sanity checking.

**Comment by josevalim** [comment]
Ok, with `a=rx` it still boots just fine. I have updated the code snippet above. I tried it both on releases/0.1.0 as well as the whole releases directory.

**Comment by goodsoft** [comment]
@josevalim `chmod` isn't recursive by default, please add `-R` option there.

```
$ chmod a=rx -R _build/dev/rel/omg/releases/0.1.0
$ _build/dev/rel/omg/bin/omg start
ERROR! Could not write runtime configuration due to reason: :eacces. Please make sure ".../omg/_build/dev/rel/omg/tmp/omg-0.1.0-20190911182215-0e20.runtime.config" is writable and accessible or choose a different path
{"init terminating in do_boot",{#{'__exception__'=>true,'__struct__'=>'Elixir.RuntimeError',message=><<"Could not write runtime configuration due to reason: :eacces. Please make sure \".../omg/_build/dev/rel/omg/tmp/omg-0.1.0-20190911182215-0e20.runtime.config\" is writable and accessible or choose a different path">>},[{'Elixir.Config.Provider',abort,1,[{file,"lib/config/provider.ex"},{line,247}]},{'Elixir.Config.Provider',boot,3,[{file,"lib/config/provider.ex"},{line,163}]},{init,eval_script,2,[]},{init,do_boot,3,[]}]}}
init terminating in do_boot ({,[{Elixir.Config.Provider,abort,1,[{_},{_}]},{Elixir.Config.Provider,boot,3,[{_},{_}]},{init,eval_script,2,[]},{init,do_boot,3,[]}]})

Crash dump is being written to: erl_crash.dump...done
```

In this case it tries to create a file in the read-only release folder though.
You can try adding
```
export RELEASE_TMP=`pwd`
```
before starting release, so that it tries to create a file in the current (writable) dir.

**Comment by josevalim** [comment]
@goodsoft generally speaking, do you think calling `chmod +w` on the file would be enough to fix it? Or are there some gotchas we should be aware of?

**Comment by josevalim** [comment]
Closing in favor of #9338. We can continue the discussion there!

**Comment by chulkilee** [comment]
Probably #9339 ?

**Comment by josevalim** [comment]
Yes, thank you! 😅

**Comment by m-roll** [comment]
Sorry for digging up an old issue, but it seems like this is still occurring for me on 1.13.3. Is there a new workaround for this?

Pretty much the exact scenario, I'm attempting to run my release after it is installed by Nix.
